part/txn.go

```diff
@@ -157,101 +157,166 @@ func (txn *Txn[T]) cloneNode(n *header[T]) *header[T] {
+			// Exact match, clone and update the value.
+			oldValue = this.getLeaf().value
+			hadOld = true
+			this = txn.cloneNode(this)
+			*thisp = this
+			this.getLeaf().value = value
+		} else {
+			// Partially matching prefix.
 			newNode := &node4[T]{
-				header: header[T]{prefix: newPrefix},
+				header: header[T]{prefix: common},
 			}
 			newNode.setKind(nodeKind4)
 
+			// Make a shallow copy of the leaf. But keep its watch channel
+			// intact since we're only manipulating its prefix.
+			oldLeafCopy := *this.getLeaf()
+			oldLeaf := &oldLeafCopy
+			oldLeaf.prefix = oldLeaf.prefix[len(common):]
+			key = key[len(common):]
+			newLeaf := newLeaf(txn.opts, key, fullKey, value)
+
+			// Insert the two leaves into the node we created. If one has
+			// a key that is a subset of the other, then we can insert them
+			// as a leaf of the node4, otherwise they become children.
 			switch {
+			case len(oldLeaf.prefix) == 0:
+				oldLeaf.prefix = common
+				newNode.setLeaf(oldLeaf)
+				newNode.children[0] = newLeaf.self()
+				newNode.keys[0] = newLeaf.prefix[0]
+				newNode.setSize(1)
+
 			case len(key) == 0:
+				newLeaf.prefix = common
 				newNode.setLeaf(newLeaf)
-				newNode.children[0] = this
-				newNode.keys[0] = this.prefix[0]
+				newNode.children[0] = oldLeaf.self()
+				newNode.keys[0] = oldLeaf.prefix[0]
 				newNode.setSize(1)
 
-			case this.prefix[0] < key[0]:
-				newNode.children[0] = this
-				newNode.keys[0] = this.prefix[0]
+			case oldLeaf.prefix[0] < key[0]:
+				newNode.children[0] = oldLeaf.self()
+				newNode.keys[0] = oldLeaf.prefix[0]
 				newNode.children[1] = newLeaf.self()
 				newNode.keys[1] = key[0]
 				newNode.setSize(2)
+
 			default:
 				newNode.children[0] = newLeaf.self()
 				newNode.keys[0] = key[0]
-				newNode.children[1] = this
-				newNode.keys[1] = this.prefix[0]
+				newNode.children[1] = oldLeaf.self()
+				newNode.keys[1] = oldLeaf.prefix[0]
 				newNode.setSize(2)
 			}
-			if !txn.opts.rootOnlyWatch || this == newRoot {
-				newNode.header.watch = make(chan struct{})
-			}
 			*thisp = newNode.self()
-			return
 		}
+	case len(key) == 0:
+		// Exact match, but not a leaf node
+		this = txn.cloneNode(this)
+		*thisp = this
+		if leaf := this.getLeaf(); leaf != nil {
+			// Replace the existing leaf
+			oldValue = leaf.value
+			hadOld = true
+			leaf = txn.cloneNode(leaf.self()).getLeaf()
+			leaf.value = value
+			this.setLeaf(leaf)
+		} else {
+			// Set the leaf
+			this.setLeaf(newLeaf(txn.opts, this.prefix, fullKey, value))
+		}
+
+	default:
+		// Partially matching prefix, non-leaf node.
+		common := commonPrefix(key, this.prefix)
+
+		this = txn.cloneNode(this)
+		*thisp = this
+		this.prefix = this.prefix[len(common):]
+		key = key[len(common):]
+
+		newLeaf := newLeaf(txn.opts, key, fullKey, value)
+		newNode := &node4[T]{
+			header: header[T]{prefix: common},
+		}
+		newNode.setKind(nodeKind4)
+
+		switch {
+		case len(key) == 0:
+			newLeaf.prefix = common
```

Comment:
Why does the newLeaf need this prefix set? `newNodes`'s prefix is `common` already, so won't this "duplicate" the common part?

Reply:
So that we can move it around and can rely on the prefix being correct. I guess we could just have it as nil and special-case when a leaf node is moved, but that's probably less easy to follow.

Reply:
Specifically this line wouldn't work: https://github.com/cilium/statedb/pull/23/files/754d4c402b1bb3f9dee86e8f953d06660081641d#diff-243d2f909d6df11ff7564c096de7f3fc5c336450a15def5587d8121b2fa86c6eR227. We'd need to carry the prefix over from the node pointing to it.